modules/finality-verifier/src/lib.rs

```diff
@@ -126,18 +126,9 @@ decl_module! {
 				<Error<T>>::InvalidAncestryProof
 			);
 
-			// If for whatever reason we are unable to fully import headers and the corresponding
-			// finality proof we want to avoid writing to the base pallet storage
-			use frame_support::storage::{with_transaction, TransactionOutcome};
-			with_transaction(|| {
-				for header in ancestry_proof {
-					if T::HeaderChain::import_header(header).is_err() {
-						return TransactionOutcome::Rollback(Err(<Error<T>>::FailedToWriteHeader))
-					}
-				}
-
-				TransactionOutcome::Commit(Ok(()))
-			})?;
+			// Note that this won't work if we ever change the `ancestry_proof` format to be
+			// sparse since this expects a contiguous set of finalized headers.
+			T::HeaderChain::append_finalized_chain(ancestry_proof);
```

Comment:
We've lost the transactional storage writes here, but we assume that the function is infallible (afaict it only panics in case of internal inconsistency)?

Reply:
Yeah, there's really no reason why this should fail given our assumptions